discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,266 @@
+package zookeeper
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/discovery"
+	"github.com/joyent/containerpilot/utils"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+func init() {
+	discovery.RegisterBackend("zookeeper", ConfigHook)
+}
+
+// ZooKeeper wraps a ZooKeeper connection handler. It also stores the
+// prefix under which ContainerPilot nodes will be registered.
+type ZooKeeper struct {
+	Client    *zk.Conn
+	Prefix    string
+	eventChan chan zk.Event
+}
+
+// ServiceNode is the serializable form of a ZooKeeper service record
+type ServiceNode struct {
+	ID      string   `json:"id"`
+	Name    string   `json:"name"`
+	Address string   `json:"address"`
+	Port    int      `json:"port"`
+	Tags    []string `json:"tags"`
+}
+
+type zookeeperRawConfig struct {
+	Endpoints interface{} `mapstructure:"endpoints"`
+	Prefix    string      `mapstructure:"prefix"`
+}
+
+func (conn ZooKeeper) ttlHandler(service *discovery.ServiceDefinition, key string) {
+	for {
+		select {
+		case ev := <-conn.eventChan:
+			_, _, _, err := conn.Client.GetW(ev.Path)
+			if err != nil {
+				log.Warning(err)
+				conn.Deregister(service)
+				return
+			}
+		case <-time.After(time.Duration(service.TTL) * time.Second):
+			conn.Deregister(service)
+			return
+		}
+	}
+}
+
+func (conn ZooKeeper) eventChanCallBack(ev zk.Event) {
+	if ev.Type == zk.EventNodeDataChanged {
+		go func() {
+			conn.eventChan <- ev
+		}()
+	}
+}
+
+// ConfigHook is the hook to register with the ZooKeeper backend
+func ConfigHook(raw interface{}) (discovery.ServiceBackend, error) {
+	return NewZooKeeperConfig(raw)
+}
+
+func connection(addresses []string, cb zk.EventCallback) (*zk.Conn, error) {
+	c, _, err := zk.Connect(addresses, time.Second, zk.WithEventCallback(cb))
+	return c, err
+}
+
+// NewZooKeeperConfig creates a new service discovery backend for zookeeper
+func NewZooKeeperConfig(raw interface{}) (*ZooKeeper, error) {
+	zookeeper := &ZooKeeper{
+		Prefix:    "/containerpilot",
+		eventChan: make(chan zk.Event),
+	}
+	config := &struct {
+		Address string `mapstructure:"address"`
+	}{}
+	if err := utils.DecodeRaw(raw, &config); err != nil {
+		return nil, err
+	}
+	conn, err := connection([]string{config.Address}, zookeeper.eventChanCallBack)
+	if err != nil {
+		return nil, err
+	}
+	zookeeper.Client = conn
+	return zookeeper, nil
+}
+
+// Deregister removes this instance from the registry
+func (conn *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
+	key := conn.getNodeKey(service)
+	exists, _, err := conn.Client.Exists(key)
```

Comment:
Right but it's racy between the check and the `Delete`. If the error is meaningful anyway (which it must be if we're `log.Warnf` it), it's better to just try to `Delete` and then catch and log that error.

```go
func (conn *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
	key := conn.getNodeKey(service)
	log.Warnf("Deregistering %s", key)
	if err := onn.Client.Delete(key, -1); err != nil {
		log.Errorf("Error on deregistering %s: %s, key, err)
	}
```

Also, you may want to consider returning an error from this function and then having the caller log it. This would make this a bit more testable.